### ISSUE TITLE
Scrolling with keyup/keydown events rather than OS keypress repetition

### ISSUE THREAD

**Issue by The-Compiler** [qutebrowser/qutebrowser#4768]
Vimium in Chrome scrolls when holding down <kbd>j</kbd> without having an initial and repeat delay caused by the key repetition.

Some users would like the same in qutebrowser - though I'm not really sure if there's a good clean implementation, as it violates the current "one key press is mapped to one command" model.

**Comment by jgkamat** [comment]
A related issue is that when enabling smooth scrolling in #3691, holding down
the key prevents scrolling from actually occurring as the re-scroll can fire
before any progress has been made. I think we can solve that case by simply
throttling the scroll events when smooth scrolling is enabled though (that is
how vimium does it, I think).

**Comment by nsklaus** [comment]
i experience quite some lag too when scrolling using keyboard on qutebrowser.
i tested with firefox and falkon, that doesn't happen with them.
i made a small video (28mb) to show the problem you can see it [here](https://ufile.io/2on1k5dv).

my desktop is i3-gaps, my term is "st", i login in tty, no DM, then i just startx.
my xinitrc is small:
```
#!/bin/sh
setbg &			        # set the background with the `setbg` script
xcompmgr &		        # xcompmgr for transparency
dunst &			        # dunst for notifications
xset r rate 300 50 & 	# Speed xrate up
unclutter &		        # Remove mouse when idle
exec i3
```
i tested adding this to my config.py, but it didn't help
```
config.bind('up', 'scroll-px 0 1')
config.bind('down', 'scroll-px 0 -1')
```
i also tested the 
`:set qt.force_software_rendering  ` with all strings, i didn't help either.
i don't understand why on same machine same web engine there is such a difference in scrolling responsivity between falkon and qutebrowser.

**Comment by The-Compiler** [comment]
Mirror of that video file: https://qutebrowser.org/tmp/4768-nsklaus.mkv

**Comment by jgkamat** [comment]
nsklaus writes:

> i experience quite some lag too when scrolling using keyboard on 
> qutebrowser.
> i tested with firefox and falkon, that doesn't happen with them.
> i made a small video (28mb) to show the problem you can see it 
> [here](https://ufile.io/2on1k5dv).

Can you try removing the scrolling widget? I think this is a duplicate of

http://www.github.com/qutebrowser/qutebrowser/issues/4475

You can also try applying 
http://www.github.com/qutebrowser/qutebrowser/pull/4498

> config.bind('up', 'scroll-px 0 1')
> config.bind('down', 'scroll-px 0 -1')

Just to be sure, you're talking about mouse scrolling here? If you are 
referring to scrolling when pressing keys, the problem is probably instead 
high overhead when processing keys.

See http://www.github.com/qutebrowser/qutebrowser/issues/4721, although this 
is not entirely fixable.

**Comment by nsklaus** [comment]
> Can you try removing the scrolling widget? 

i'm sorry i have no idea what you're talking about here .. what "scrolling widget" ?

> Just to be sure, you're talking about mouse scrolling here? 

short version:
no, i'm not talking about mouse scrolling. obviously not.
please, consider the title of this issue, and the first line of my previous post.
i have provided two videos, and a comment to explain.

longer version:
so let me rephrase it, in an attempt to be even clearer about it.
the issue i'm having, is that when i press "arrow up key" , or "arrow down key" the scroll result is laggy, slow. i experience even small freezes where qutebrower have to wait 2 or 3 seconds to perform the scroll. please watch the videos again. 
look at how slow the response from qutebrowser is when i scroll with arrow up and arrow down.
so you can compare easily, i show me, performing the same attempt on falkon browser (that is: scrolling with arrow keys up/down) and the result is good, it's smooth, there is no lag, no freeze.
in these two video i also show, for comparison purpose that, when i scroll with the mouse in qutebrowser, it is smooth and good. so my question is:
why mouse scrolling is smooth and good, while keyboard arrow scrolling is bad and slow.
especialy considering it doesn't happen on falkon which uses the same engine as qutebrowser.

**Comment by jgkamat** [comment]
nsklaus writes:

> i'm sorry i have no idea what you're talking about here .. what "scrolling 
> widget" ?

The scroll widget in statusbar.widgets

> no, i'm not talking about mouse scrolling. obviously not.
> please, consider the title of this issue, and the first line of my previous 
> post.
> i have provided two videos, and a comment to explain.

I only see a single video, where's the second one? I don't see how this issue 
(avoiding the initial repeat delay) is relevant to your problem as I 
understand it (lag when scrolling continuously).

>> i experience quite some lag too when scrolling using keyboard on 
>> qutebrowser.
>> i tested with firefox and falkon, that doesn't happen with them.
>> i made a small video (28mb) to show the problem you can see it 
>> [here](https://ufile.io/2on1k5dv).

To me, all the scrolling (mouse and keyboard) in that video is much more 
laggy/slow than falkon, and most of it does not involve any keypresses 
(start->1:17), so it's difficult to tell what you're talking about.

> longer version:
> so let me rephrase it, in an attempt to be even clearer about it.
> the issue i'm having, is that when i press "arrow up key" , or "arrow down 
> key" the scroll result is laggy, slow. i experience even small freezes where 
> qutebrower have to wait 2 or 3 seconds to perform the scroll. please watch 
> the videos again.

FWIW, I can't reproduce your issue with the keyboard, until I get below a 
repeat rate of 30, at which point falkon chokes as well. qutebrowser is still 
much worse than falkon at smooth scrolling though, but I don't think we can 
fix that without c extensions (if at all). Can you try in qutebrowser -T, and 
if that dosen't help, qutebrowser -T with 
http://www.gitlab.com/jgkamat/qutebrowser/tree/integration ?


Here's what I see: https://www.youtube.com/watch?v=XfPeAR6DlbQ

**Comment by The-Compiler** [comment]
A somewhat related idea from the [mailing list](https://lists.schokokeks.org/pipermail/qutebrowser/2020-January/000691.html): an auto-hidden tab-bar could be visible as long as shift is held down when switching tabs with shift-j/shift-k.

**Comment by toniz4** [comment]
I would find this helpful, i use a high repetition rate, and because that (+ a potato computer), the scrolling jitters a little sometimes. That could be made optional thru a config, and then it would be a good idea to implement another parameter to change the scrolling speed without having to change the repetition rate. I wish i knew how to code in python so i could help you out with that ;(

Thanks for making a awesome browser!

**Comment by The-Compiler** [comment]
From [Reddit](https://www.reddit.com/r/qutebrowser/comments/j3s3fq/is_vimiumlike_scrolling_possible_in_qb/g7e9g82/?context=3) by @bovine3dom:

> I thought quite a bit about adding keyup binds to Tridactyl here - https://github.com/tridactyl/tridactyl/pull/1645 - the idea being that you'd bind keydown to start scrolling and keyup to stop scrolling.
>
> I quite liked the idea that it would let you introduce layers to Tridactyl - e.g. holding z would put you in a different mode with different binds until you let go of z. I only stopped because I introduced a bug and then forgot about the PR.

- Some thoughts about keyup binds are in #5654 
- Custom modes (for those layers) in #5755 
- This would likely also introduce auto-scroll at the same time, if there was a `:scroll-start <speed>` binding on keydown and a `:scroll-end` binding on keyup.

**Comment by xav-ie** [comment]
I think a reasonable workaround or fix could be this qutebrowser script:
```js
// ==UserScript==
// @name         Smooth Scroll Injector
// @version      1.0
// @description  Injects smooth scroll CSS to webpages
// @author       Your Name
// @match        *://*/*
// @grant        none
// ==/UserScript==

(function() {
    'use strict';

    const css = 'html { scroll-behavior: smooth !important; }';
    const style = document.createElement('style');
    style.appendChild(document.createTextNode(css));
    document.head.appendChild(style);
})();
```

I also turn on the smooth scroll setting. I am happy to see that scrolling is not jittery and the behavior is identical to Vimium. This was the one thing keeping me from using this amazing browser in the past so hopefully this helps some of yall.

**Comment by Combowombo** [comment]
thanks for the suggestion xav-ie. It has made my scrolling smooth pretty consistently. The only thing I'm trying to get is continuous scroll when "d" is held down. I asked Claude to see if i can solve for this via my qutebrowser config and it gave me the following code:

import qutebrowser
from qutebrowser.api import interceptor, key_module

# Define a flag to track if the "d" key is held down
is_d_pressed = False

# Define a function to handle key presses
@interceptor.define(key_module.KeyPress)
def handle_key_press(info):
    global is_d_pressed

    # Check if the "d" key is pressed
    if info.key.lower() == 'd':
        is_d_pressed = True
        # Execute the scrolling function while the key is held down
        while is_d_pressed:
            qutebrowser.scroll.scroll('dx', 0, 50)  # Scroll down by 50 pixels
            qutebrowser.utils.later(50, handle_key_press, info)  # Call the function again after 50ms

# Define a function to handle key releases
@interceptor.define(key_module.KeyRelease)
def handle_key_release(info):
    global is_d_pressed

    # Check if the "d" key is released
    if info.key.lower() == 'd':
        is_d_pressed = False

Unfortunately, I think it hallucinated the "key_module" library/package but from reading it it seems to be on the right track. Anyone have any thoughts on how to further correct?

**Comment by The-Compiler** [comment]
Everything in that answer is completely hallucinated.

**Comment by Walter-o** [comment]
Is an experimental toggle for this in the plans?

If this truly violates a rule deemed sacred to some people then maybe add it as an experimental option that remains opt-in.

**Comment by The-Compiler** [comment]
"Just add a toggle" isn't the answer here. Someone will have to think about how to best add something that should be bound on key-up (both from the code and from the user's perspective).

**Comment by bovine3dom** [comment]
In Tridactyl I was going to try using U and D as modifiers to add to keys to specify keyup/keydown as we do for e.g. holding shift. I also introduced the concept of an optional modifier `?` that meant a key could be triggered or not.

Then internally any keys that didn't have explicit U/D modifiers were translated e.g. `xy` to `<D-x><U?-x><D-y><U?-y>`

But then I hit some boring bug where I couldn't bind to spacebar and I got bored :)
